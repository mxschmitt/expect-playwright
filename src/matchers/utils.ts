import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
