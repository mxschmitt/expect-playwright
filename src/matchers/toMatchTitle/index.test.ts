import { assertSnapshot } from "../tests/utils"

describe("toMatchTitle", () => {
  afterEach(() => page.setContent(""))

  describe("with string argument", () => {
    it("positive", async () => {
      await page.setContent("<title>Foo</title>")
      await expect(page).toMatchTitle("Foo")
    })

    it("negative", async () => {
      await page.setContent("<title>Foobar</title>")
      await assertSnapshot(() => expect(page).toMatchTitle("Foo"))
    })

    describe("with 'not' usage", () => {
      it("positive", async () => {
        await page.setContent("<title>Foobar</title>")
        await expect(page).not.toMatchTitle("Foo")
      })

      it("negative", async () => {
        await page.setContent("<title>Foo</title>")
        await assertSnapshot(() => expect(page).not.toMatchTitle("Foo"))
      })
    })
  })

  describe("with regex argument", () => {
    it("positive", async () => {
      await page.setContent("<title>Foobar</title>")
      await expect(page).toMatchTitle(/foo/i)
    })

    it("negative", async () => {
      await page.setContent("<title>Foobar</title>")
      await assertSnapshot(() => expect(page).toMatchTitle(/Bar/))
    })

    describe("with 'not' usage", () => {
      it("positive", async () => {
        await page.setContent("<title>Foo</title>")
        await expect(page).not.toMatchTitle(/bar/)
      })

      it("negative", async () => {
        await page.setContent("<title>Foobar</title>")
        await assertSnapshot(() => expect(page).not.toMatchTitle(/Foo/))
      })
    })
  })
})
