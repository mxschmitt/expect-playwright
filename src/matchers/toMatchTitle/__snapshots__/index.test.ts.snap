// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchTitle with regex argument negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchTitle[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m/Bar/[39m
Received: [31m\\"Foobar\\"[39m"
`;

exports[`toMatchTitle with regex argument with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoMatchTitle[2m([22m[32mexpected[39m[2m)[22m

Expected: not [32m/Foo/[39m"
`;

exports[`toMatchTitle with string argument negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchTitle[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"Foo\\"[39m
Received: [31m\\"Foo[7mbar[27m\\"[39m"
`;

exports[`toMatchTitle with string argument with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoMatchTitle[2m([22m[32mexpected[39m[2m)[22m

Expected: not [32m\\"Foo\\"[39m"
`;
